rfd/0083-license-expiration-warnings.md

```diff
@@ -0,0 +1,155 @@
+---
+authors: edwarddowling (edward.dowling@goteleport.com), r0mant (roman@goteleport.com)
+---
+
+
+# RFD 83 - License warnings
+
+## Required approvers
+
+- Engineering: `@r0mant`
+- Product: `@klizentas && @xinding33`
+- Security: `@reedloden`
+
+## What
+
+This RFD proposes a way to display license warnings to self hosted users both before and after license expiration.
+
+## Why
+
+To encourage the renewal of licenses after expiration and let users know before their license expires.
+
+## Success criteria
+
+Teleport displays license warnings to users in both CLI and web UI starting from a certain time prior to the license expiration, as well as post-expiration.
+
+## Scope
+
+Self-hosted Teleport will be the target of this RFD. Teleport Cloud is out of scope because the Cloud license does not have proper expiration yet. When Cloud implements proper license expiration, this feature will kick in automatically.
+License warning should be displayed in both Web UI and CLI (tsh/tctl).
+
+### CLI UX
+
+License warnings should be displayed in CLI (tsh/tctl) 
+TSH - on “tsh login”, in “tsh status”
```

Comment:
`tsh status` currently makes no network calls, using only the local filesystem to generate its output.  This can be very nice for folks with slower networks, since it gives them a fast way to get key information at a glance.

When Implementing version alerts I opted not to show them in `tsh status` for this reason.  IMO if license warnings are to be displayed in `tsh status`, they should be behind a flag (e.g. `-v/--verbose`) in order to preserve the snappiness of the current `tsh status` impl by default.

Reply:
Not including these alerts in the standard `tsh status` makes sense to me. There is a comment in tsh.go mentioning the possibility of adding a `tsh alerts ls` command eventually which we could use instead of the `tsh status -v` as an option for viewing alerts.

Although when the license has expired we would like to show the warnings on every command. We could exclude the `tsh status` command from this to keep it snappy. What are your thoughts on this @xinding33

Reply:
Let's see what @xinding33 says but I'm pretty sure we do want to show warnings on `tsh status` by default, esp. if your license has expired. We can cache the alerts on the client somehow?

Reply:
Thinking more about this, maybe caching on the client is not a great idea since we'll run into problem with invalidation if someone has updated the license while you're logged in, so might get unnecessarily complicated.

@fspmarshall I'm not sure tbh we'll be able to maintain this properly that `tsh status` doesn't make network calls to the cluster. There are alerts that we definitely do want to show on status (e.g. these license ones). Hiding them behind `-v` flag is I think not much better than not showing them at all as I'm sure people don't run verbose status very often and we do want these alerts to be seen.

@xinding33 It's a product call at the end of the day so let us know what you think.

Reply:
My perspective is that we need this check in `tsh status` (and not put it behind verbose flag) when tsh is connecting to an enterprise cluster to support our pricing model. Latency is either acceptable or unacceptable. In the case that it's acceptable, adding it to `tsh status` won't be detrimental to use experience and in the case that it's unacceptable, we need to fix it because even if we don't add it to `tsh status`, high latency will impact other use cases. What's the expected added latency to `tsh status` based on the current suggested implementation? Typically, anything less than 100ms is not noticeable.

Reply:
We could add a timeout to the request so that in the event the request takes longer than say 350ms it times out and dumps the status as normal. In this case the status output could have a little note that says: 

```
license check timed out, run: the following to check your license information:
tsh check-license
```

This way we get the best of both worlds.

Reply:
@fspmarshall  @xinding33 are both of you satisfied with the timeout approach? If so i can update the design doc and add it into the implementation i'm working on.

Reply:
Timeout approach seems reasonable to me.  How about instead of querying specifically for license alerts, we just query for all *high severity* alerts.  If we're going to show alerts on `tsh status` we might as well show security patch alerts too.

Reply:
This makes sense to me. @EdwardDowling Let's update the RFD to address these two things:

- High severity alerts will be shown on tsh status.
- tsh status will use some reasonable timeout (say, 300-500ms) for fetching alerts.

And merge.

Reply:
@rosstimothy @fspmarshall I hear you guys and in general I agree with the sentiment. On the other hand, I'm not really sure what specific problem adding an extra call would introduce here, besides increased latency which in the worst-case scenario will make the status command feel not instant but still snappy. Where does the requirement of not having `tsh status` make calls to the cluster come from? I think it's just a happenstance that it doesn't make them now and wasn't necessarily a conscious design decision (unless I'm mistaken).

We could in theory fetch all alerts on login and cache them locally, but then we'd run into problems with invalidation, etc which sounds like over-engineering for something as simple as showing alerts so I'd like to avoid this approach.

If we want to be able to show key information without waiting on network, we can also consider displaying current "local" status right away, while we're fetching alerts in the background. Then the command will exit after it's received/displayed alerts (or upon timeout). Not sure if it's better or worse UX though.

WDYT?

Reply:
So is the current plan to only show the alert on login and after every tctl once the license expires? Leaving the tsh commands untouched for?

Reply:
@r0mant @rosstimothy  could confirm the most recent changes to the RFD are what you were thinking I can go ahead and merge this?